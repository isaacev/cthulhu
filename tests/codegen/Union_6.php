<?php

namespace Union_6 {
  abstract class Result {}
  class Left extends \Union_6\Result {
    function __construct($a) {
      $this->{0} = $a;
    }
  }
  class Right extends \Union_6\Result {
    function __construct($a) {
      $this->{0} = $a;
    }
  }
  function main() {
    $a = \mt_rand(0, 5);
    if ($a == 0) {
      new \Union_6\Left(true);
    } else if ($a == 1) {
      new \Union_6\Left(true);
    } else if ($a == 2) {
      new \Union_6\Left(false);
    } else {
      new \Union_6\Right("unknown integer");
    }
    return null;
  }
}

namespace {
  \Union_6\main(null);
}
