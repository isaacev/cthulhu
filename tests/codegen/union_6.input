type Result('a, 'b) =
  | Left('a)
  | Right('b);

fn attempt(success: Bool, val: 'a) -> Result('a, Str) {
  if success {
    Result::Left(val)
  } else {
    Result::Right("unable to generate a true value")
  }
}

#[entry]
fn main() -> () {
  let r: Result(Bool, Str) = match ::Random::integer(0, 5) {
    0 => Result::Left(true),
    1 => Result::Left(true),
    2 => Result::Left(false),
    _ => Result::Right("unknown integer"),
  };
}
