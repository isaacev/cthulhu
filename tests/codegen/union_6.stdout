<?php

namespace Kernel\Types {
  abstract class Maybe {}

  class Just extends \Kernel\Types\Maybe {
    function __construct($a) {
      $this->{0} = $a;
    }
  }

  class None extends \Kernel\Types\Maybe {}
}

namespace Kernel\Builtins {
  // #[inline]
  function mt_rand($a, $b) {
    return \mt_rand($a, $b);
  }
}

namespace Random {
  // #[inline]
  function integer($min, $max) {
    return \Kernel\Builtins\mt_rand($min, $max);
  }
}

namespace union_6 {
  abstract class Result {}

  class Left extends \union_6\Result {
    function __construct($a) {
      $this->{0} = $a;
    }
  }

  class Right extends \union_6\Result {
    function __construct($a) {
      $this->{0} = $a;
    }
  }

  // #[entry]
  function main() {
    $b = \Random\integer(0, 5);
    if ($b == 0) {
      $a = new \union_6\Left(true);
    } else if ($b == 1) {
      $a = new \union_6\Left(true);
    } else if ($b == 2) {
      $a = new \union_6\Left(false);
    } else if (true) {
      $a = new \union_6\Right("unknown integer");
    }
    $r = $a;
  }
}

namespace {
  \union_6\main();
}
